src/ballet/bmtree/fd_bmtree.inc

```diff
@@ -0,0 +1,153 @@
+#include "fd_bmtree.h"
+
+#define BMTREE_(x)     FD_EXPAND_THEN_CONCAT4(fd_bmtree,BMTREE_VARIANT,_,x)
+#define BMTREEU_(x)    FD_EXPAND_THEN_CONCAT4(FD_BMTREE,BMTREE_VARIANT,_,x)
+#define BMTREE_NODE    BMTREE_(node_t)
+#define BMTREE_COMMIT  BMTREE_(commit_t)
+#define BMTREE_NODE_SZ BMTREEU_(NODE_SZ)
+
+void
+BMTREE_(commit_init)( BMTREE_COMMIT * state,
+                      ulong           leaf_cnt ) {
+  state->leaf_idx = 0UL;
+  state->leaf_cnt = leaf_cnt;
+}
+
+/* fd_bmtree*_merge: Derives a single branch node.
+
+   Computes `SHA-256([0x01] | A | B)` and writes the result into out.
+   Arbitrary pointer aliasing between out, A, B is permitted. */
+static void
+BMTREE_(merge)( BMTREE_NODE       * out,
+                BMTREE_NODE const * A,
+                BMTREE_NODE const * B ) {
+  fd_sha256_t sha;
+  fd_sha256_init( &sha );
+
+  /* Prepend static one-byte prefix identifying a branch node. */
+  uchar const prefix[1] = { FD_BMTREE_PREFIX_BRANCH };
+  fd_sha256_append( &sha, &prefix, 1UL );
+  /* Write tree nodes. */
+  fd_sha256_append( &sha, A, BMTREE_NODE_SZ );
+  fd_sha256_append( &sha, B, BMTREE_NODE_SZ );
+
+  /* Write 32-byte hash result into temporary buffer. */
+  uchar hash[FD_SHA256_HASH_SZ] __attribute__((aligned(FD_SHA256_ALIGN)));
+  fd_sha256_fini( &sha, hash );
+
+  /* Write tree node into out, discarding last few bytes. */
+  fd_memcpy( out, hash, BMTREE_NODE_SZ );
+}
+
+/* fd_bmtree*_commit_insert: Accumulates a single leaf node into the tree.
+
+   Updates `state->leaf_idx++`, U.B. if value >= `state->leaf_cnt`.
+
+   Maintains the invariant that the left node of the last node pair
+   for each layer is copied to `state->node_buf`.
+
+   This serves to allow the algorithm to derive a new parent branch node
+   for any pair of children, once the (previously missing) right node
+   becomes available. */
+static void
+BMTREE_(commit_insert)( BMTREE_COMMIT *     state,
+                        BMTREE_NODE const * leaf ) {
+  BMTREE_NODE tmp;
+  fd_memcpy( &tmp, leaf, sizeof(tmp) );
+
+  /* Walk the tree upwards from the bottom layer.
+
+     `tmp` contains a previously missing right node
+     which is used to derive a branch node,
+     together with the previously buffered value in `node_buf`.
+
+     Each iteration, merges that pair of nodes into a new branch node.
+     Terminates if the new branch node is the left node of a pair. */
+
+  ulong layer, cursor;
+  for(
+    /* `layer` starts at 0 (leaf nodes) and increments each iteration.
+       `cursor` is the number of known nodes in the current layer. */
+    layer=0UL, cursor=++state->leaf_idx;
+    /* Continue while the right node in the last pair is available.
+       Loop always terminates as `cursor` eventually reaches `1`. */
```

Comment:
I read "eventually reaches 1" as eventually it is true that cursor==1, which I think is false. If cursor starts at 6, it'll end at 3.

``Loop always terminates as `cursor` is always strictly positive and thus has at least one bit set to `1`. */``